### PR TITLE
Result (lightweight Try)

### PR DESCRIPTION
**Pull request by fwbrasil**
Following a similar approach as https://github.com/getkyo/kyo/pull/464. The implementation can avoid allocations in case of success but has to allocate in case of failures.

### CODE REVIEW

File: kyo-core/shared/src/main/scala/kyo/Result.scala [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+        def unapply[T](self: Result[T]): Option[T] =
+            (self: @unchecked) match
+                case _: Failure[?] => None
+                case s: SuccessFailure[?] =>
+                    Some(s.unnest.asInstanceOf[T])
+                case s: T @unchecked => Some(s)
+
+    end Success
+
+    case class Failure[+T](exception: Throwable)
```
Comment (hearnadam) [inline kyo-core/shared/src/main/scala/kyo/Result.scala:49]: Assuming this will be the backing for `Fiber`, do we want to store other information? Trace? Do we need a `Cause` type like ZIO?

Comment (fwbrasil) [reply kyo-core/shared/src/main/scala/kyo/Result.scala:49]: Yes, I'm planning to use it to avoid `T < IOs` in fiber completions. The `Option`/`Maybe` was a little detour to explore how to support nesting without allocations :)

I think we could keep the trace information in the exception itself since we can decide which exceptions benefit from a origin `Trace` and which don't.

File: kyo-core/shared/src/main/scala/kyo/Result.scala [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+            (self: @unchecked) match
+                case Success(value) => ev(value)
+                case failure        => failure.asInstanceOf[Result[U]]
+
+        inline def map[U](inline f: T => U): Result[U] =
+            (self: @unchecked) match
+                case Success(value) => Result(f(value))
+                case failure        => failure.asInstanceOf[Result[U]]
+
+        inline def fold[U](inline ifFailure: Throwable => U, inline ifSuccess: T => U): U =
```
Comment (hearnadam) [inline kyo-core/shared/src/main/scala/kyo/Result.scala:98]: Would it be beneficial to use multiple parameter lists to improve type inference?

File: kyo-core/shared/src/test/scala/kyoTest/ResultTest.scala [hunk 3]
```diff
@@ -0,0 +1,285 @@
...
+        }
+    }
+
+    "edge cases" - {
+        "should handle exceptions during fold" in {
+            val tryy      = Success(1)
+            val exception = new RuntimeException("exception")
+            val result =
+                try
+                    tryy.fold(_ => throw exception, _ => throw exception)
```
Comment (hearnadam) [inline kyo-core/shared/src/test/scala/kyoTest/ResultTest.scala:253]: https://github.com/scala/scala/blob/2.13.x/src/library/scala/util/Try.scala#L288-L289

I don't think this is the behavior most users of `Try` would expect.

Comment (fwbrasil) [reply kyo-core/shared/src/test/scala/kyoTest/ResultTest.scala:253]: good catch! fixed

File: kyo-core/shared/src/test/scala/kyoTest/ResultTest.scala [hunk 4]
```diff
@@ -0,0 +1,301 @@
...
+    "getOrElse" - {
+        "returns the value for Success" in {
+            assert(Success(1).getOrElse(0) == 1)
+        }
+        "returns the default value for Failure" in {
+            assert(Failure[Int](new Exception("error")).getOrElse(0) == 0)
+        }
+    }
+
+    "orElse" - {
```
Comment (hearnadam) [inline kyo-core/shared/src/test/scala/kyoTest/ResultTest.scala:55]: Can you add a type inference test:

https://github.com/zio/zio/issues/8890

ZIO.succeed(Option.empty[List[String]]).someOrElse(List.empty)
```
Result(Option.empty[List[String]).orElse(Result(List.empty))
```

**Review by hearnadam — APPROVED** [review]

File: kyo-core/shared/src/test/scala/kyoTest/ResultTest.scala (hunk 4 above)
Comment (fwbrasil) [reply kyo-core/shared/src/test/scala/kyoTest/ResultTest.scala:55]: It doesn't compile. What return type should it have? `Result[IterableOnce]`?

Comment (hearnadam) [reply kyo-core/shared/src/test/scala/kyoTest/ResultTest.scala:55]: Ah no, sorry that is a bad example. I forgot the `someOrElse` operator will squash the Option.

Perhaps just `Result(List.empty[String]).orElse(List.empty)`

Comment (fwbrasil) [reply kyo-core/shared/src/test/scala/kyoTest/ResultTest.scala:59]: I've added a couple of inference tests. It seems to be working correctly. I think inference here is simpler than in ZIO's case.
